### DISCUSSION TITLE
Questions with 'ControlSurfacePidController'

### DISCUSSION THREAD

**Discussion by Alanpwu** [ImperialCollegeLondon/sharpy#133] [Q&A] [answered]
Hi, 
I have a flying wing system with 1362 states and 893 inputs, two of the inputs are control surface deflection and their time derivative, the rests are accounted for gust. I want to design a controller (u = -Kx) only for surface deflections and their time derivatives. And my questions are: 
1. Does the 'ControlSurfacePidController' calculate the for all 893 inputs or there is a way to make it only calculate the control surface deflection and its time derivative? 
2. Does the 'ControlSurfacePidController' solver internally takes the states values as input or the inputs needed to specified? 
3. In 'ControlSurfacePidController' doc the gains are float type, what if my K is a matrix?

**Comment by ngoiz** [comment, marked as answer]
The controller you are referring to is intended to be used in the time domain simulation of the nonlinear SHARPy (for single-input single-output systems).

For linear state-space systems I would recommend exporting the state-space (A, B, C, D) and working externally to SHARPy (in Matlab or Python) and implementing your linear controller there.

**Reply by Alanpwu** [reply]
Thank you for your reply, it's really helpful. 

Another question is that in the example of 'Flutter Analysis of a Goland Wing using the SHARPy Linear Solver', the number of input of the linearised aeroelastic system is always (No. of structural states)*1.5 + 2*(No. of control surface). I am wandering what these structural inputs represent? 

And it is common that the state-space A matrix is rank deficient?

**Reply by ngoiz** [reply]
I replied to the follow-up in #134 

About the A matrix being rank deficient, it depends on the system you are trying to assemble. For free-flying aircraft it may be the case if you include integrals of the rigid body degrees of freedom as part of the states (and will appear as eigenvalues at `0 + 0j`)
